### ISSUE TITLE
initial lithostatic pressure in spherical geometries

### ISSUE THREAD

**Issue by mfmweerdesteijn** [geodynamics/aspect#4715]
When applying initial lithostatic pressure profiles on far boundaries (to make them 'open') in spherical geometries (here 2D chunk), large strain rate anomalies occur at these boundaries, resulting in unwanted free surface deformation (if used in combination with free surface, as is the case here). Here are plots of the free surface through time (colors) for a linearly increasing load with time at the center of the domain. Ignore the plot titles.

**1: boundary traction and closed lateral boundaries**
![bt_cb](https://user-images.githubusercontent.com/57097877/169574179-ee283bff-52fd-475a-bf3c-dc7eea2888ff.png)

**2: boundary traction and open lateral boundaries**
![bt_ob](https://user-images.githubusercontent.com/57097877/169574189-5e0c58a7-c760-47be-bcd8-99e4afe052cd.png)

**2: strain rates for case 2**
<img width="541" alt="bt_ob_strainrate" src="https://user-images.githubusercontent.com/57097877/169574837-22e034e7-9b36-4822-8ba7-9e71e704a81b.png">

**3: no boundary traction, only open lateral boundaries**
![nbt_ob](https://user-images.githubusercontent.com/57097877/169574202-576e3cf1-56ab-4fa1-bb88-1b28b67620c2.png)

**3: the latter but zoomed in**
![nbt_ob_zoom](https://user-images.githubusercontent.com/57097877/169574251-236d9dd2-eabf-4911-966c-521c72056fbc.png)

The initial lithostatic pressure causes strain rate anomalies at the applied boundaries, in spherical geometries. These strain rate anomalies can be of significant size w.r.t. the boundary traction. Is this a mapping problem? Any ideas?

**Comment by naliboff** [comment]
Thanks @mfmweerdesteijn! So, it looks like a small amount of deformation is occurring at the boundaries even when no surface traction is applied (i.e., the model should be static). Is this interpretation correct? 

If so, can you add an image of the strain rates for the "open boundaries, no surface tractions" case? If there are strain anomalies, can you add an image of the pressure field and adjust the color scale to highlight small differences at a specific depth? 

Thanks!

**Comment by mfmweerdesteijn** [comment]
Indeed, there are deformations even when no surface traction is applied, so coming from the initial lithostatic pressure at the lateral boundaries applied.

open boundaries, no surface tractions strain rate images:
START (t=0)
<img width="448" alt="nbt_ob_strainrate_start" src="https://user-images.githubusercontent.com/57097877/169606288-5e3e51d2-00c7-4b9b-8388-bde2862a7d74.png">

END (t=100 yr)
<img width="454" alt="nbt_ob_strainrate_end" src="https://user-images.githubusercontent.com/57097877/169606303-aa3749af-5d0e-40e2-869f-da597b324074.png">

I played around with the pressure field but didn't manage to visualize small differences.
But I'm doing a test now with equal resolution throughout the domain, because it seems like some of the high strain rates are located near the border of the mesh resolution.

<img width="411" alt="mesh" src="https://user-images.githubusercontent.com/57097877/169606805-e480daf4-9954-4aed-8fde-8a82409306a2.png">

**Comment by mfmweerdesteijn** [comment]
No boundary traction, only initial lithostatic pressure on both lateral boundaries, equal mesh resolution throughout domain: again strain rate anomaly, at bottom boundary, resulting in unwanted free surface deformation. 
<img width="449" alt="nbt_ob_eqmesh" src="https://user-images.githubusercontent.com/57097877/169816657-ee650ff0-a918-482c-8377-2cf9ce9728f5.png">
![nbt_ob_eqmesh_sd](https://user-images.githubusercontent.com/57097877/169816673-ce350c2a-4e14-4971-bb17-de440c31ee5b.png)

**Comment by anne-glerum** [comment]
Hi @mfmweerdesteijn , I went through the initial_lithostatic_pressure plugin, and I didn't see anything obvious why the computed pressure wouldn't be correct in the spherical case. The chunk geometry does have its issues though. Do you see the same anomalies in a quarter spherical shell? Also, there are some checks in the initial_lithostatic_pressure plugin that are only checked when run in Debug mode. Have you done that?

**Comment by mfmweerdesteijn** [comment]
When I use a spherical shell with 90 degrees opening angle and with initial lithostatic pressure on the lateral boundaries, I get the error: "After displacement of the mesh, this function can no longer be used to determine whether a point lies in the domain or not." This is from the point_is_in_domain function. But I don't see why it would work for other geometries (box and chunk) with mesh deformation enabled. I then changed the conditions of the AssertThrow in the spherical shell to the same conditions as in the chunk.

from:

    template <int dim>
    bool
    SphericalShell<dim>::point_is_in_domain(const Point<dim> &point) const
    {
      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
                  this->get_timestep_number() == 0,
                  ExcMessage("After displacement of the mesh, this function can no longer be used to determine whether a point lies in the domain or not."));

to (based on conditions in the point_is_in_domain function from the chunk geometry):

    template <int dim>
    bool
    SphericalShell<dim>::point_is_in_domain(const Point<dim> &point) const
    {
      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
                  this->get_timestep_number() == 0 ||
                  this->get_timestep_number() == numbers::invalid_unsigned_int,
                  ExcMessage("After displacement of the mesh, this function can no longer be used to determine whether a point lies in the domain or not."));

These are btw the conditions in the box geometry:

    template <int dim>
    bool
    Box<dim>::point_is_in_domain(const Point<dim> &point) const
    {
      AssertThrow(!this->get_parameters().mesh_deformation_enabled ||
                  this->simulator_is_past_initialization() == false,
                  ExcMessage("After displacement of the free surface, this function can no longer be used to determine whether a point lies in the domain or not."));

Why isn't the code consistent?

With this change, the model runs, and unfortunately I still get strain rate anomalies at the bottom boundary at the lateral boundaries, and at a few random along the bottom boundary:

surface deformation:
![sphericalshell](https://user-images.githubusercontent.com/57097877/170265119-59c5ff80-12db-405b-8b5e-896e9eda756d.png)

strain rate (one capture of the 90 degrees shell with strain rate anomaly in the corner at the lateral boundary, and an anomaly randomly placed along the bottom boundary):
<img width="391" alt="sphericalshell_sr" src="https://user-images.githubusercontent.com/57097877/170265213-4886f6e9-f068-41c0-a05f-404fab114613.PNG">

**Comment by anne-glerum** [comment]
Hi @mfmweerdesteijn thank you for the additional info. 

> Why isn't the code consistent?
These functions have ben improved on over time and have apparently not been updated everywhere. The last implementation (for the box) is the most correct. It would be great if you can put in a PR or make an issue to remind us to unify the functions.

The strain rate anomalies in the bottom corners: Couldn't they derive from the difference in velocity directions? The bottom boundary is tangential, but velocities along the open boundaries can have tangential as well as normal components (I think). So at the corner, there could be a change in direction, leading to a higher strain rate. Can you check what the velocity vectors look like there? 

Also, is there flow into or out of the domain (on both open boundaries)? Then the question is, why is the computed pressure not good enough and induces flow? What are the velocity magnitudes?

The surface deviations are very small (less than 1 m right), especially compared to the model resolution (what is the resolution?). Maybe they become smaller with higher resolution. By default a 1000 integration points are used to compute the lithostatic profile. How many elements are there in the depth direction?

You can also analytically compute the lithostatic pressure at the bottom of your model and compare it with what is computed in the plugin (with a print statement). When the density is constant, this is just density*gravity*depth.
